Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+        m_stringBuilder.append("compute");
+        break;
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::LocationAttribute& location)
+{
+    m_stringBuilder.append("[[attribute(", location.location(), ")]]");
+}
+
+void FunctionDefinitionWriter::visit(AST::TypeDecl& type)
+{
+    AST::Visitor::visit(type);
+}
+
+void FunctionDefinitionWriter::visit(AST::ArrayType& type)
+{
+    ASSERT(type.maybeElementType());
+    ASSERT(type.maybeElementCount());
+    m_stringBuilder.append("array<");
+    visit(*type.maybeElementType());
+    m_stringBuilder.append(", ");
+    visit(*type.maybeElementCount());
+    m_stringBuilder.append(">");
+}
+
+void FunctionDefinitionWriter::visit(AST::NamedType& type)
+{
+    if (type.name() == "i32"_s)
+        m_stringBuilder.append("int");
+    else if (type.name() == "f32"_s)
+        m_stringBuilder.append("float");
+    else if (type.name() == "u32"_s)
+        m_stringBuilder.append("unsigned");
+    else
+        m_stringBuilder.append(type.name());
+}
+
+void FunctionDefinitionWriter::visit(AST::ParameterizedType& type)
+{
+    switch (type.base()) {
+    case AST::ParameterizedType::Base::Vec2:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 2>");
+        break;
+    case AST::ParameterizedType::Base::Vec3:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 3>");
+        break;
+    case AST::ParameterizedType::Base::Vec4:
+        m_stringBuilder.append("vec<");
+        visit(type.elementType());
+        m_stringBuilder.append(", 4>");
+        break;
+    case AST::ParameterizedType::Base::Mat2x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat2x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat2x4:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat3x4:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x2:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x3:
+        // TODO
+        break;
+    case AST::ParameterizedType::Base::Mat4x4:
+        // TODO
+        break;
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Parameter& parameter)
+{
+    visit(parameter.type());
+    m_stringBuilder.append(" ", parameter.name());
+    for (auto& attribute : parameter.attributes()) {
+        m_stringBuilder.append(" ");
+        checkErrorAndVisit(attribute);
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Expression& expression)
+{
+    AST::Visitor::visit(expression);
+}
+
+void FunctionDefinitionWriter::visit(AST::CallableExpression& call)
+{
+    bool first = true;
+    if (call.target().kind() == AST::Node::Kind::ArrayType) {
```

Comment:
There's not a more elegant way of handling this? it seems weird for code to indicate we're trying to call an array.